README.md
# jbrowse_elasticsearch

[![Build Status](https://travis-ci.org/elsiklab/jbrowse_elasticsearch.svg?branch=master)](https://travis-ci.org/elsiklab/jbrowse_elasticsearch)

A JBrowse add-on for indexing names with elasticsearch. Allows searching full text descriptions of genes!

## Pre-requisites

- elasticsearch
- nodejs/npm
- cpanm

## Installation

Run the setup script

    bash setup.sh

Then load your tracks

    flatfile-to-json.pl --nameAttributes note,id,description,name --gff docs/tutorial/data_files/volvox.gff3 --trackLabel test --trackType CanvasFeatures

And then load the tracks into elasticsearch

    bin/generate-elastic-search.pl

Then add the plugin to JBrowse by adding something like this to trackList.json or `jbrowse_conf.json`

    "plugins": ["ElasticSearch"]

Finally start the helper app (starts app.js as middleware for elasticsearch queries)

    npm start


## Loading full text descriptions

If you have a feature such as

    chr23  RefSeq  gene    2475803 2809862 .   -   .   ID=gene28777;Name=514682;Dbxref=NCBI_Gene:514682,BGD:BT30338;symbol_ncbi=PRIM2;description=primase%2C DNA%2C polypeptide 2 (58kDa);gene_synonym=PRIM2A;feature_type=Protein Coding


Then running

    flatfile-to-json.pl --trackLabel RefSeq --gff file.gff --nameAttributes symbol_ncbi,gene_synonym,description,dbxref

This would make `symbol_ncbi` the "primary key" and associate the `gene_synonym`, `description`, and `dbxref` as "descriptions" of that gene (the search box doesn't distinguish the field type, they all just become descriptions)



 
## Screenshot

![](img/example.png)


## Configuration


- elasticSearchUrl - the location of the express.js middleware, e.g. http://localhost:3000
- elasticIndexName - optional: the name of a specific elastic index that you loaded data into, normally used if you have multiple genomes

These are automatically added to trackList.json by the bin/generate-elastic-search.pl

## Troubleshooting


* The setup.sh installs the jbrowse libs to the system with cpanm, if there are problems with setup.sh make sure cpanm is installed correctly

* For the gene descriptions to be indexed, the `--nameAttributes` flag should be used with flatfile-to-json.pl with something like `--nameAttributes name,description,other_field_to_index`. By default --nameAttributes is id,name,alias. The order of arguments in --nameAttributes is important, the first should be the unique name, id or symbol (or whatever is acceptable to appear in the name box in the popup). Other arguments after the first are then all associated with the first as descriptions, with multiple descriptions being allowed

* If your tracklist significantly changes you may need to regenerate your names index (the elasticsearch data refers to indexes in the tracks and if your tracklist changes then you will not be referring to the correct track, jbrowse default generate-names.pl does this also)

* If your descriptions are on the "mRNA" features in your GFF but you load at the "gene" level then your descriptions will not be indexed. You can create an extra track loaded at the "mRNA" level using --type mRNA in flatfile-to-json.pl and then re-run the generate-elastic-search pipeline. 


## Multiple genomes configuration

Use --genome argument to bin/generate-elastic-search.pl which creates different indexes (the elasticsearch equivalent of a different database) for each genome that your run


## Middleware configuration

Normally, you can start the express.js middleware for jbrowse_elasticsearch (a small service that queries elasticsearch for you) by running "npm install" and "npm start" in the root folder of this repo. In a production configuration, you may wish to use a reverse proxy to make this accessible from a standard http endpoint rather than opening up port 3000 to the public. Therefore put something like this basic config in your apache config file

    ProxyPass /elastic http://localhost:3000
    ProxyPassReverse /elastic http://localhost:3000

Read up on reverse proxies before doing so. Forward proxies are dangerous and you do not want to enable this, but a configuration like the above is safe. You can also, instead of relying on "npm start" to keep the service running, run it on Passenger Phusion

    Alias /elasticsearch /mnt/webdata/jbrowse/plugins/ElasticSearch
    <Location /elasticsearch>
        PassengerBaseURI /elasticsearch
        PassengerAppRoot /mnt/webdata/jbrowse/plugins/ElasticSearch

        PassengerAppType node
        PassengerStartupFile bin/www
    </Location>

Then point the elasticSearchUrl parameter to http://yoursite/elasticsearch

Note that the express.js middleware is pretty small, but it is recommended to use an endpoint like this rather than expose the elasticsearch REST API to the public.

## Defaults

* http://localhost:3000 is the default express.js port, can be overridden in app.js and in --url param to generate-elastic-search.pl

* http://localhost:9200 is the default elasticsearch port, can be overridden in app.js and in --elasticurl param to generate-elastic-search.pl


## Feedback

Feel free to provide feedback, my first foray into elasticsearch!



ChangeLog.md
# Version 1.0.5

- Pin to specific bioperl version

# Version 1.0.4

- Add "exact" search
- Fix for for super-new dgrid versions (post 1.12.3 in jbrowse)
- Add note about using passenger phusion

# Version 1.0.3

- Add total hits to dialog box
- Add ability to support multiple genomes via different elasticsearch indexes (see --genome argument to generate-elastic-search.pl)

# Version 1.0.2

- Cleanup some source code and tests

# Version 1.0.1

- Add ChangeLog.md

# Version 1.0.0

- Added custom dialog box to handle full text searching
- Added ability to only use prefix searching, which helps for the REST Names API autocompletion purposes
- Added scripts to automatically load existing data directory into elastic search database


src/perl5/Bio/JBrowse/Cmd/ElasticSearch.pm
package Bio::JBrowse::Cmd::ElasticSearch;

=head1 NAME

Bio::JBrowse::Cmd::ElasticSearch - generate-elastic-search.pl.

=cut

use strict;
use warnings;


use base 'Bio::JBrowse::Cmd';
use Search::Elasticsearch ();
use List::Util ();

use GenomeDB ();

sub option_defaults {(
    dir => 'data',
    completionLimit => 20,
    locationLimit => 100,
    url => 'http://localhost:3000/',
    elasticurl => 'http://localhost:9200',
    genome => '',
    mem => 256 * 2**20,
    tracks => [],
)}

sub option_definitions {(
    "dir|out=s",
    "verbose|v+",
    "elasticurl=s",
    "genome=s",
    "url=s",
    "help|h|?",
    'tracks=s@'
)}

sub initialize {
    my ( $self ) = @_;
    $self->{e} = Search::Elasticsearch->new(
        nodes => $self->opt('elasticurl')
    );
}

sub run {
    my ( $self ) = @_;

    my $outDir = $self->opt('dir');
    -d $outDir or die "Output directory '$outDir' does not exist.\n";
    -w $outDir or die "Output directory '$outDir' is not writable.\n";
    my $gdb = GenomeDB->new( $outDir );

    my $refSeqs = $gdb->refSeqs;
    unless( @$refSeqs ) {
        die "No reference sequences defined in configuration, nothing to do.\n";
    }
    my @tracks = grep $self->track_is_included( $_->{label} ),
                      @{ $gdb->trackList || [] };
    unless( @tracks ) {
        die "No tracks. Nothing to do.\n";
    }

    # find the names files we will be working with
    my $names_files = $self->find_names_files( \@tracks, $refSeqs );
    unless( @$names_files ) {
        warn "WARNING: No feature names found for indexing,"
             ." only reference sequence names will be indexed.\n";
    }

    # store the list of tracks that have names
    $self->load( $refSeqs, $names_files );


    # set up the name store in the trackList.json
    $gdb->modifyTrackList( sub {
                               my ( $data ) = @_;
                               $data->{elasticIndexName} = $self->opt('genome');
                               $data->{elasticSearchUrl} = $self->opt('url');
                               return $data;
                           });
    return;
}

sub load {
    my ( $self, $ref_seqs, $names_files ) = @_;

    # convert the stream of name records into a stream of operations to do
    # on the data in the hash store
    my $op_stream = $self->make_operation_stream( $self->make_name_record_stream( $ref_seqs, $names_files ), $names_files );

    # hash each operation and write it to a log file
    my $bulk = $self->{e}->bulk_helper(
        index   => 'gene' . $self->opt('genome'),
        type    => 'loc',
        verbose => $self->opt('verbose')
    );
    my $counter=0;
    while ( my $op = $op_stream->() ) {
        my ( $lc_name, $op_name, $record ) = @$op;

        # not allowed to index names with '.'
        if($lc_name ne '.') {
            $bulk->index({ source => {
                description => $record->[0],
                name => $record->[2],
                track_index => $self->{stats}{tracksWithNames}[$record->[1] || 0],
                ref => $record->[3],
                start => $record->[4],
                end => $record->[5]
            }});
        }
    }
    $bulk->flush;

}

sub make_file_record {
    my ( $self, $track, $file ) = @_;
    -f $file or die "$file not found\n";
    -r $file or die "$file not readable\n";
    my $gzipped = $file =~ /\.(txt|json|g)z$/;
    my $type = $file =~ /\.txtz?$/      ? 'txt'  :
               $file =~ /\.jsonz?$/     ? 'json' :
               $file =~ /\.vcf(\.gz)?$/ ? 'vcf'  :
               $file =~ /\.gff(\.gz)?$/ ? 'gff'  :
                                           undef;

    if( $type ) {
        return { gzipped => $gzipped, fullpath => $file, type => $type, trackName => $track->{label} };
    }
    return;
}

sub track_is_included {
    my ( $self, $trackname ) = @_;
    my $included = $self->{included_track_names} ||= do {
        my @tracks = @{ $self->opt('tracks') };
        my $inc = { map { $_ => 1 } map { split ',', $_ } @tracks };
        @tracks ? sub { $inc->{ shift() } } : sub { 1 };
    };
    return $included->( $trackname );
}


my $OP_ADD_EXACT  = 1;

sub make_operations {
    my ( $self, $record ) = @_;

    my $lc_name = lc $record->[0];
    unless( defined $lc_name ) {
        unless( $self->{already_warned_about_blank_name_records} ) {
            warn "WARNING: some blank name records found, skipping.\n";
            $self->{already_warned_about_blank_name_records} = 1;
        }
        return;
    }

    my @ops = ( [ $lc_name, $OP_ADD_EXACT, $record ] );

    $self->{stats}{operations_made} += scalar @ops;

    return @ops;
}


sub make_name_record_stream {
    my ( $self, $refseqs, $names_files ) = @_;
    my @names_files = @$names_files;

    my $name_records_iterator = sub {};
    my @namerecord_buffer;

    # insert a name record for all of the reference sequences
    for my $ref ( @$refseqs ) {
        $self->{stats}{name_input_records}++;
        $self->{stats}{namerecs_buffered}++;
        my $rec = [ @{$ref}{ qw/ name length name seqDir start end seqChunkSize/ }];
        $self->{stats}{total_namerec_bytes} += length join(",",$rec);
        push @namerecord_buffer, $rec;
    }


    my %trackHash;
    my $trackNum = 0;



    return sub {
        while( ! @namerecord_buffer ) {
            my $nameinfo = $name_records_iterator->() || do {
                my $file = shift @names_files;
                return unless $file;
                $name_records_iterator = $self->make_names_iterator( $file );
                $name_records_iterator->();
            } or return;
            my @aliases = map { ref($_) ? @$_ : $_ }  @{$nameinfo->[0]};
            foreach my $alias ( @aliases ) {
                my $track = $nameinfo->[1];
                unless ( defined $trackHash{$track} ) {
                    $trackHash{$track} = $trackNum++;
                    push @{$self->{stats}{tracksWithNames}}, $track;
                }
                $self->{stats}{namerecs_buffered}++;
                push @namerecord_buffer, [
                    $alias,
                    $trackHash{$track},
                    @{$nameinfo}[2..$#{$nameinfo}]
                ];
            }
        }
        return shift @namerecord_buffer;
    };
}




sub find_names_files {
    my ( $self, $tracks, $refseqs ) = @_;

    my @files;
    for my $track (@$tracks) {
        for my $ref (@$refseqs) {
            my $dir = File::Spec->catdir($self->opt('dir'), "tracks", $track->{label}, $ref->{name});

            # read either names.txt or names.json files
            my $name_records_iterator;
            my $names_txt  = File::Spec->catfile( $dir, 'names.txt'  );
            if( -f $names_txt ) {
                push @files, $self->make_file_record( $track, $names_txt );
            }
            else {
                my $names_json = File::Spec->catfile( $dir, 'names.json' );
                if( -f $names_json ) {
                    push @files, $self->make_file_record( $track, $names_json );
                }
            }
        }

        # try to detect VCF tracks and index their VCF files
        if( $track->{storeClass} && ( $track->{urlTemplate} && $track->{urlTemplate} =~ /\.vcf\.gz/ || $track->{storeClass} =~ /VCFTabix$/ ) ) {
            my $path = File::Spec->catfile( $self->opt('dir'), $track->{urlTemplate} );
            if( -r $path ) {
                push @files, $self->make_file_record( $track, $path );
            }
            else {
                warn "VCF file '$path' not found, or not readable.  Skipping.\n";
            }
        }

    }

    return \@files;
}

sub make_operation_stream {
    my ( $self, $record_stream, $names_files ) = @_;
    my @operation_buffer;
    return sub {
        unless( @operation_buffer ) {
            if( my $name_record = $record_stream->() ) {
                push @operation_buffer, $self->make_operations( $name_record );
            }
        }
        return shift @operation_buffer;
    };
}

# each of these takes an input filename and returns a subroutine that
# returns name records until there are no more, for either names.txt
# files or old-style names.json files
sub make_names_iterator {
    my ( $self, $file_record ) = @_;
    if( $file_record->{type} eq 'txt' ) {
        my $input_fh = $self->open_names_file( $file_record );
        # read the input json partly with low-level parsing so that we
        # can parse incrementally from the filehandle.  names list
        # files can be very big.
        return sub {
            my $t = <$input_fh>;
            if( $t ) {
                $self->{stats}{name_input_records}++;
                $self->{stats}{total_namerec_bytes} += length $t;
                return eval { JSON::from_json( $t ) };
            }
            return undef;
        };
    }
    elsif( $file_record->{type} eq 'json' ) {
        # read old-style names.json files all from memory
        my $input_fh = $self->open_names_file( $file_record );

        my $data = JSON::from_json(do {
            local $/;
            my $text = scalar <$input_fh>;
            $self->{stats}{total_namerec_bytes} += length $text;
            $text;
        });

        $self->{stats}{name_input_records} += scalar @$data;

        return sub { shift @$data };
    }
    elsif( $file_record->{type} eq 'vcf' ) {
        my $input_fh = $self->open_names_file( $file_record );
        no warnings 'uninitialized';
        return sub {
            my $line;
            while( ($line = <$input_fh>) =~ /^#/ ) {}
            return unless $line;

            $self->{stats}{name_input_records}++;
            $self->{stats}{total_namerec_bytes} += length $line;

            my ( $ref, $start, $name, $basevar ) = split "\t", $line, 5;
            $start--;
            return [[$name],$file_record->{trackName},$name,$ref, $start, $start+length($basevar)];
        };
    }
    else {
        warn "ignoring names file $file_record->{fullpath}.  unknown type $file_record->{type}.\n";
        return sub {};
    }
}

sub open_names_file {
    my ( $self, $filerec ) = @_;
    my $infile = $filerec->{fullpath};
    if( $filerec->{gzipped} ) {
        # can't use PerlIO::gzip, it truncates bgzipped files
        my $z;
        eval {
            require IO::Uncompress::Gunzip;
            $z = IO::Uncompress::Gunzip->new( $filerec->{fullpath }, -MultiStream => 1 )
                or die "IO::Uncompress::Gunzip failed: $IO::Uncompress::Gunzip::GunzipError\n";
        };
        if( $@ ) {
            # fall back to use gzip command if available
            if( `which gunzip` ) {
                open my $fh, '-|', 'gzip', '-dc', $filerec->{fullpath}
                   or die "$! running gunzip";
                return $fh;
            } else {
                die "cannot uncompress $filerec->{fullpath}, could not use either IO::Uncompress::Gunzip or gzip";
            }
        }
        else {
            return $z;
        }
    }
    else {
        open my $fh, '<', $infile or die "$! reading $infile";
        return $fh;
    }
}


1;
